### PR TITLE
Feat/residual load

### PR DESCRIPTION
**Pull request by fwiedemair**
closes https://github.com/AGGM-AG/pypsa-at-planning/issues/142

## Changes proposed in this Pull Request
Adds visuals for residual load and residual load duration curve. The residual load is calculated as the sum of all electrical loads/withdrawal at the AC bus excluding transmisison and storage (losses from the latter are counted towards the load9 minus the generation from all fluctuating renewables (solar, wind, ror).  List of changes:

- Implementation of a statistics loss  function that calculates the loss for a bus_carrier.
- Implementation of a residual load visual
- Implementation of a residual load duration curve visual

## Checklist

**Required:**
- [x] Changes are tested locally and behave as expected.
- [x] Code and workflow changes are documented.
- [x] A brief description of the changes has been added to `Changelog.md`.
- [x] All Sourcery Bot review suggestions have been implemented or discarded with an explanation.
- [ ] All github actions succeed.
- [x] Manual claude review has been executed and al points adressed or implemented

**If applicable:**
- [ ] For new data sources or versions, [these instructions](https://pypsa-eur.readthedocs.io/en/latest/data_sources) have been followed.
- [ ] New rules are documented in the appropriate `docs-at/` files.

## Summary by Sourcery

Add residual load evaluation and visualization, including statistics for network losses and duration curves.

New Features:
- Provide residual load time series and residual load duration curve views configurable via evals config, including plotting options.
- Add a duration curve transformation utility to convert time series data into duration-curve format for plotting.

Enhancements:
- Introduce a loss statistics method to attribute branch losses to sending bus locations and expose it through ESMStatistics.
- Extend timeseries plotting and export infrastructure with per-view overrides, optional duration-curve handling, and custom styling for residual load plots.
- Fix unit scaling in time-series scaling utility and wire residual load views into the evals view registry.

Tests:
- Add unit tests for the new loss statistic behavior and the duration-curve conversion helper.

### CODE REVIEW

**Comment by fwiedemair** [comment]
Ad sourcery review:
- True, fixed with slightly more complicated function
- Typo fixed
- That was an unintended commit. Fixed.

**Comment by fwiedemair** [comment]
Local claude review:

# Review: `feat/residual_load` vs `main`

## 🔴 Critical bug: `ESMStatistics.loss()` uses an invalid unit string

`evals/statistic.py:1004`
```python
df.attrs["unit"] = "MWh_el"
```
Every other statistic in this file uses `"MWh"` (see `evals/statistic.py:306,382,475`, etc.), and
`"MWh_el"` is **not a key** in `constants.UNITS`. The `_el`/`_th`/`_LHV` suffix convention is only
stripped from the *target* unit inside `scale()` (`evals/utils.py:363-365`), never from the
*source* unit read out of `df.attrs["unit"]`. As a result:

1. `Exporter.__init__` (`evals/fileio.py:304-306`) builds
   `units = {stat.attrs["unit"] for stat in statistics}` from
   `[load, link_withdrawal, fluctuating_generation, losses]` in `simple_residual_load`
   (`evals/views/common.py:575-579`). Since `load`/`link_withdrawal`/`fluctuating_generation` are
   `"MWh"` and `losses` is `"MWh_el"`, this immediately raises
   `ValueError("Mixed units cannot be exported: {'MWh', 'MWh_el'}")`.
2. Even if that check were bypassed, `scale()` would raise `KeyError: 'MWh_el'` at
   `UNITS[is_unit]` (`evals/utils.py:369`).

Both failures were reproduced directly:

```python
# Exporter mixed-units check
units = {s.attrs["unit"] for s in [a, b]}  # {"MWh", "MWh_el"} -> len != 1 -> raises

# scale() KeyError
scale(df, "GW")  # KeyError: 'MWh_el'
```

**Both new views (`view_timeseries_residual_load`, `view_residual_load_duration_curve`) will
crash at runtime.** This isn't caught by the added unit tests because `test_statistic.py` only
calls `.loss()` directly and never routes the result through `Exporter`.

**Fix:** change `df.attrs["unit"] = "MWh_el"` to `"MWh"` to match the `withdrawal`/`supply`
statistics it's combined with.

## 🟡 Minor issues

1. **`evals/views/common.py:497` — unused `bus_carrier` variable.**
   `_parse_view_config_items` returns a config-driven `bus_carrier`, but `simple_residual_load`
   ignores it and hardcodes `bus_carrier="AC"` in all four `collect_myopic_statistics` calls.
   Works today only because the TOML configs happen to set `bus_carrier = "AC"`; if that config
   value ever changes, it will be silently ignored. Prefer using the parsed `bus_carrier`
   variable, or drop it from `_parse_view_config_items`'s return tuple for this call site.

2. **`evals/utils.py:487` — `plot_df_out.name = "Duration (hours)"` is dead code.**
   `DataFrame` has no built-in meaning for `.name` (unlike `Series.name`), and nothing in
   `evals/plots/timeseries.py` reads `df.name`. The actual x-axis label comes from the
   `xaxis_title` `global_override` in `config.default.toml`. This line does nothing and should
   probably be removed, or if intended to set the column axis label it should be
   `plot_df_out.columns.name = ...`.

3. **`evals/fileio.py:477-480` — removed hardcoded `pivot_index`/`xaxis_title` defaults for
   `ESMTimeSeriesChart`.** Previously every `ESMTimeSeriesChart` view got `xaxis_title=""` and
   `pivot_index=[YEAR, LOCATION, CARRIER]` automatically. Now this is only set via each view's
   `global_override` in `config.default.toml`. All 4 existing timeseries views were correctly
   migrated, but any future `ESMTimeSeriesChart` view that forgets to add a `global_override`
   block will silently fall back to `build_plot_config`'s default `pivot_index`
   (`[year, location, carrier, bus_carrier]` — note the extra `bus_carrier` level) rather than
   failing loudly. Worth a code comment or validation to prevent silent misconfiguration.

4. **Naming clarity** — in `simple_residual_load`, the variable `load_carriers`
   (`evals/views/common.py:510-526`) actually represents "AC withdrawal carriers that are not
   transmission/storage losses" (i.e. non-`Load`-component withdrawal like heat pumps, EVs,
   electrolysis). Combined with `load`, this correctly implements "all withdrawal except
   transmission and storage" per the docstring, but the name is easy to misread as "carriers
   belonging to the Load component." A name like `other_withdrawal_carriers` would be clearer.

## ✅ Things that check out

- `_port_location` and the direction-aware `loss_func` logic (assign `p0+p1` to whichever port
  has non-negative flow) is correct — verified against both new unit tests
  (`test_loss_attributes_to_sending_bus_location`, `test_loss_with_line_flow_reversing_direction`)
  by hand-tracing the math, and the tests pass.
- `to_duration_curve` correctly sorts descending and computes cumulative duration in hours
  (verified against `test_to_duration_curve`, including leap-year edge case).
- `scale()`'s bugfix (`is_unit / to_unit` → `UNITS[is_unit] / UNITS[to_unit]`) is a correct,
  necessary fix (previously would fail dividing two strings).
- Doc updates (view count 23→25, new `loss` statistic row, new view table rows,
  `global_override` config key) are accurate and match the code.
- `ruff check` passes clean; all existing + new tests pass.

## Recommendation

Fix the `"MWh_el"` unit string (item 1) before merging — this is a blocking functional bug that
breaks the very feature this PR introduces. The other points are lower-priority cleanups.

**Comment by fwiedemair** [comment]
Ad claude review:

- The critical issue is not a bug. The code runs as intended. As for discrepancieswith units for timeseries data these will be handled in https://github.com/AGGM-AG/pypsa-at-planning/issues/265

Minor issues:
1. Good points. replaced hard-coded "AC" values with variable
2. Removed dead code
3. Noted. Will be considered in future implementations.
4. Also noted. No change necessary for now

File: evals/views/common.py [hunk 1]
```diff
@@ -470,6 +470,115 @@ def simple_storage_capacity(
...
+        - loss_carriers
+    )
+
+    load = (
+        collect_myopic_statistics(
+            nc,
+            statistic="withdrawal",
+            bus_carrier=bus_carrier,
+            aggregate_time=False,
+            aggregate_components=None,
```
Comment (pworschischek-aggmag) [inline evals/views/common.py]: I think you can simply specify `components="Load"` and drop the filters below as well as the `aggreagte_components=None`. If just one component is given, the index level should be dropped by default if I rememeber correctly

File: evals/views/common.py [hunk 2]
```diff
@@ -470,6 +470,115 @@ def simple_storage_capacity(
...
+
+    link_withdrawal = (
+        collect_myopic_statistics(
+            nc,
+            statistic="withdrawal",
+            bus_carrier=bus_carrier,
+            aggregate_time=False,
+            aggregate_components=None,
+        )
+        .pipe(filter_by, component=list(withdrawal_comps), carrier=list(load_carriers))
```
Comment (pworschischek-aggmag) [inline evals/views/common.py]: its possible to request components and carrier filtering directly via pypsa.statistics API

File: evals/views/common.py [hunk 3]
```diff
@@ -470,6 +470,115 @@ def simple_storage_capacity(
...
+    fluctuating_generation = (
+        collect_myopic_statistics(
+            nc,
+            statistic="supply",
+            bus_carrier=bus_carrier,
+            aggregate_time=False,
+            aggregate_components=None,
+        )
+        .pipe(filter_by, carrier=fluctuating_carriers)
+        .mul(-1)
```
Comment (pworschischek-aggmag) [inline evals/views/common.py]: I always added comments when fliping signs. Why do we need flipped signs here?

File: evals/config.default.toml [hunk 4]
```diff
@@ -1351,6 +1355,10 @@ database_plot_type = "timeseries"
 database_bus_carrier = "hydrogen"
 database_specifier = ""
 
+[view_timeseries_hydrogen.global_override]
```
Comment (pworschischek-aggmag) [inline evals/config.default.toml:1365]: Is this something users want to change? If not, I think its better added directly in the views that require those settings. Simply update the config there and add a comment on why the update is needed

File: evals/config.default.toml [hunk 5]
```diff
@@ -1590,3 +1606,55 @@ database_bus_carrier = "mixed"
...
+
+[view_timeseries_residual_load]
+name = "Residual Load"
+unit = "GW"
+file_name = "residual_load_{location}_{year}"
+bus_carrier = "AC"  # AC bus_carrier
+fluctuating_carriers=["onwind", "solar", "solar-hsat", "offwind-dc", "offwind-ac", "ror", "solar rooftop"]
+storage_links = ["battery charger", "PHS charger"]
+chart = "ESMTimeSeriesChart"
+cutoff = 0.00001  # 1 MWh
```
Comment (pworschischek-aggmag) [inline evals/config.default.toml]: If unit is GW, shouldn't this be 0.001 for 1 MW?

File: evals/fileio.py [hunk 6]
```diff
@@ -470,13 +477,7 @@ def export(self, result_path: Path, subdir: str) -> None:
         if chart_class == plots.ESMGroupedBarChart:
             self.defaults.xaxis_title = ""
         elif chart_class == plots.ESMTimeSeriesChart:
-            self.defaults.xaxis_title = ""
```
Comment (pworschischek-aggmag) [inline evals/fileio.py:473]: Not against the removal of defaults here, they are technical dept and should be purged. I just wonder why the rLDC does not have the same index levels as time series charts?

File: evals/fileio.py [hunk 7]
```diff
@@ -543,7 +544,9 @@ def consistency_checks(self) -> None:
 
     def default_checks(self) -> None:
         """Perform integrity checks for views."""
-        if self.view_config.get("chart") == "SankeyChart":
+        if self.view_config.get("chart") == "SankeyChart" or self.view_config.get(
```
Comment (pworschischek-aggmag) [inline evals/fileio.py]: Let's discuss changing the default checks logic instead. I'd like to keep the config layer slim (fewer values). If new views should not run checks previously run by all views (= default checks), we should make checks an explicit list per view instead of adding additional config keys.

File: evals/statistic.py [hunk 8]
```diff
@@ -47,6 +47,41 @@
 logger = logging.getLogger(__file__)
 
 
+def _port_location(n: Network, c: str, port: str = "") -> pd.Series:
```
Comment (pworschischek-aggmag) [inline evals/statistic.py]: Doesn't https://github.com/AGGM-AG/pypsa-at/blob/acfda7dc3869bbc9afb69716a2e59b6650062c3b/evals/fileio.py#L50 do the same?

File: evals/statistic.py [hunk 9]
```diff
@@ -865,3 +900,106 @@ def technical_potential(  # noqa: D417
...
+
+        Returns
+        -------
+        :
+            Loss of components in the network, with the grouping
+            keys (e.g. component, location, carrier, bus_carrier) as rows and
+            snapshots as columns.
+
+        """
+        components = components if components else ["Link", "Line", "Transformer"]
```
Comment (pworschischek-aggmag) [inline evals/statistic.py]: Nit: `components = components or ["Link", "Line", "Transformer"]` does the same and is used extensively in other libraries to replace mutable default arguments

File: evals/statistic.py [hunk 10]
```diff
@@ -865,3 +900,106 @@ def technical_potential(  # noqa: D417
...
+            dynamic = n.c[c].dynamic
+            own = dynamic[f"p{port}"]
+            other = dynamic[f"p{'1' if port == '0' else '0'}"]
+            loss = (own + other).where(own >= 0, 0).clip(lower=0)
+            return loss.T
+
+        # substitute a port-aware "location" grouper (see _port_location)
+        # so that bus0/bus1 resolve to different locations, regardless
+        # of what is currently registered under the "location" name.
+        loss_groupby = groupby
```
Comment (pworschischek-aggmag) [inline evals/statistic.py]: superfluous additional label, simply use groupby in the `isinstance()` call but keep the `loss_groupby` variable (it carries the information that something changed)

File: evals/utils.py [hunk 11]
```diff
@@ -1192,3 +1194,45 @@ def get_latest_results_folder() -> Path:
...
+    (datetime) columns are replaced by the cumulative number of hours
+    they represent, so that the result can be plotted as x=duration
+    (hours), y=sorted value. Snapshots are assumed to be equidistant;
+    the duration of the last snapshot is inferred by wrapping around to
+    one year after the first snapshot.
+
+    Parameters
+    ----------
+    df
+        DataFrame with equidistant datetime snapshots as columns.
```
Comment (pworschischek-aggmag) [inline evals/utils.py:1214]: maybe raise if equidistance is not given?

File: evals/fileio.py [hunk 12]
```diff
@@ -319,12 +320,15 @@ def __init__(
...
         title = view_config["name"] + TITLE_SUFFIX
         self.defaults.title = title
+        self.defaults.name = view_config["name"]
         self.defaults.file_name_template = view_config["file_name"]
         self.defaults.cutoff = view_config["cutoff"]
         self.defaults.category_orders = view_config["legend_order"]
         self.defaults.database_plot_type = view_config["database_plot_type"]
         self.defaults.database_bus_carrier = view_config["database_bus_carrier"]
         self.defaults.database_specifier = view_config["database_specifier"]
+        if "global_override" in view_config.keys():
```
Comment (pworschischek-aggmag) [inline evals/fileio.py:277]: Is it possible to override in the view layer instead?

**Review by pworschischek-aggmag — CHANGES_REQUESTED** [review]
I'd like to get rid of the `[*.global_override]` section in the config, if possible. 

Apart from that, just some simplifications and a proposal to update the checks logic instead of introducting new configurations.

File: evals/views/common.py (hunk 1 above)
Comment (fwiedemair) [reply evals/views/common.py]: Done

File: evals/views/common.py (hunk 2 above)
Comment (fwiedemair) [reply evals/views/common.py]: Done

File: evals/views/common.py (hunk 3 above)
Comment (fwiedemair) [reply evals/views/common.py]: Added comment. Reason: residual load = load - fluctuating renewables

File: evals/config.default.toml (hunk 4 above)
Comment (fwiedemair) [reply evals/config.default.toml:1365]: Moved settings to view code.

File: evals/config.default.toml (hunk 5 above)
Comment (fwiedemair) [reply evals/config.default.toml]: Correct. Fixed.

File: evals/fileio.py (hunk 6 above)
Comment (fwiedemair) [reply evals/fileio.py:473]: rLDC has Duration in hours as x-axis

File: evals/fileio.py (hunk 12 above)
Comment (fwiedemair) [reply evals/fileio.py:277]: Moved relevant overrides to the view layer

File: evals/statistic.py (hunk 9 above)
Comment (fwiedemair) [reply evals/statistic.py]: Done

File: evals/utils.py (hunk 11 above)
Comment (fwiedemair) [reply evals/utils.py:1214]: Added raise

File: evals/statistic.py (hunk 10 above)
Comment (fwiedemair) [reply evals/statistic.py]: Changed

File: evals/statistic.py (hunk 8 above)
Comment (fwiedemair) [reply evals/statistic.py]: Good catch, yes. Moved get_location to evals/utils and used that instead

File: evals/fileio.py (hunk 7 above)
Comment (fwiedemair) [reply evals/fileio.py]: Modified to use config `skip_default_checks` in sankey as well

**Review by pworschischek-aggmag — APPROVED** [review]
Good to go from my side!
